pxr/imaging/hdSt/materialXFilter.cpp

```diff
@@ -1296,8 +1227,37 @@ _IsTopologicalShader(TfToken const& nodeId)
     return TfStringStartsWith(nodeId.GetString(), "ND_swizzle_");
 }
 
-// Build the topoNetwork, equivalent to the given hdNetwork but anonymized and 
-// stripped of non-topological parameters to better re-use the generated shader.  
+// Helper functions generating hashes which are stable from one run of the
+// application to another
+namespace
+{
+    // The default implementation
+    template <typename T>
+    void _AppendPersistentHash(Tf_HashState& hashState, T const& value)
+    {
+        TfHashAppend(hashState, value);
+    }
+
+    // TfHashAppend hashes TfTokens not as strings, but as pointers to interned
+    // strings which are not stable from run to run
+    void _AppendPersistentHash(
+        Tf_HashState& hashState, TfToken const& token)
+    {
+        TfHashAppend(hashState, token.GetString());
+    }
+
+    // A VtValue may store a TfToken, in which case we need to convert it to a
+    // string in order to avoid the same pitfall as above
+    void _AppendPersistentHash(Tf_HashState& hashState, VtValue const& vtValue)
+    {
+        if (vtValue.IsHolding<TfToken>()) {
+            TfHashAppend(hashState, vtValue.Get<TfToken>().GetString());
+        } else {
+            TfHashAppend(hashState, vtValue.GetHash());
+        }
+    }
```

Comment:
Without these workarounds, the cache always generated new, random hashes, and could never find an existing file in the cache directory.